### ISSUE TITLE
feature request: matplotlib visualize log scale

### ISSUE THREAD

**Issue by kratsg** [alexander-held/cabinetry#113]
Would be nice for data-mc to have an option to flip log scale on (hint: `ax1.set_yscale('log')` in the right spot).

**Comment by alexander-held** [comment]
Thanks for the suggestion, this is a useful feature. There are a few different ways to configure this:

- produce both linear and log scale version for all plots by default
- keyword argument in `visualize.data_MC` to switch between linear/log for all figures (maybe with an added option to only produce a subset of figures)
- config option for each region (something like `LogScale: True`) to control the behavior

So far none of these purely cosmetic settings exist in the config, but sooner or later I think they need to be added there. I would like to have some separation between config options that define the model, and others that only adjust the visualization. I am not sure whether that is possible without duplicating the configuration structure though (regions/samples/systematics).

Happy to hear your thoughts about how you would expect to configure this behavior.

**Comment by kratsg** [comment]
Maybe @henryiii can comment here. Personally, I appreciate the ROOT style of being able to apply log-scale on a plot after it's been drawn (e.g. to change the scale/redraw if necessary). One annoying (finnicky) thing is that you generally have to adjust your ranges between linear scale and log-scale usually...

The default should be to automatically determine if log scale is needed. I recall two heuristics one can use:

- are any samples > 1-2 orders of magnitude in nevents than the other samples you're drawing?
- are there any bins > 1-2 orders of magnitude than other bins you're drawing?

This could probably cover all the use cases here. Although being able to fine tune things such as y-axis ranges, and so on is probably also desirable.

**Comment by alexander-held** [comment]
> ROOT style of being able to apply log-scale on a plot after it's been drawn

Do you mean saving as `.root` and modifying that, or interactively modifying the figure while it's being drawn? A unified format to store all information needed to re-draw a plot (a `.json` or so) would be great, and I would be happy to adopt that.

The idea to automatically determine whether a log scale is needed sounds great, I think I'll start with that. Getting a good-looking y axis range might be tricky, eventually manual control is probably needed there.

**Comment by kratsg** [comment]
> Do you mean saving as `.root` and modifying that, or interactively modifying the figure while it's being drawn? A unified format to store all information needed to re-draw a plot (a `.json` or so) would be great, and I would be happy to adopt that.

It would likely be nicer to provide a set of histogram objects that can be modified before saving.



> The idea to automatically determine whether a log scale is needed sounds great, I think I'll start with that. Getting a good-looking y axis range might be tricky, eventually manual control is probably needed there.

Yeah, so I'm thinking you can provide a `log-scale=auto/true/false` where `auto` defaults, and `true/false` are to override the automatic behavior.

**Comment by alexander-held** [comment]
> It would likely be nicer to provide a set of histogram objects that can be modified before saving.

How would you want to interact with them? By receiving the histograms and writing your own plotting function? `visualize.data_MC` builds the histograms needed for the plot, and then hands them to `contrib.matplotlib_visualize`. The function `contrib.matplotlib_visualize.data_MC` receives a list of dictionaries (that contain the relevant histogram information), and an array for the total model uncertainty. That information could instead be returned to the user.

There is a `method` argument in `visualize.data_MC`, which is intended as a switch to reach other backends (which currently do not exist). I was thinking of adding a mechanism where you define your own plotting function, and then `cabinetry` uses that instead.

**Comment by kratsg** [comment]
> There is a `method` argument in `visualize.data_MC`, which is intended as a switch to reach other backends (which currently do not exist). I was thinking of adding a mechanism where you define your own plotting function, and then `cabinetry` uses that instead.

It would be nice to provide a callback hook, that gets called with every histogram made, for users to do various manipulations maybe?

**Comment by alexander-held** [comment]
#119 added automatic determination for linear/log scale. When the (expected) yield in any two bins is different by more than 2 orders of magnitude, a log scale is used. `log_scale` keyword arguments for `visualize.data_MC` and `visualize.data_MC_from_histograms` allow to override the automatic behavior. The log scale plots get saved with a `_log` suffix in the name.

> provide a callback hook

Could you show me an example of how something like this would look like?

**Comment by kratsg** [comment]
maybe something like

```python
def visualize.data_MC(...., callback=None):

  # do all stuff to make fig,ax

  if callback:
    callback(fig, ax)
```

and then one provides something like

```python
def do_something(fig, ax):
  ax.set_yscale('log')

visualize.data_MC(...., callback=do_something)
```

**Comment by alexander-held** [comment]
Thanks, this looks like a good idea. Until this is added, the latest release [0.1.4](https://github.com/alexander-held/cabinetry/releases/tag/v0.1.4) can be used for the basic log scale functionality from #119.

**Comment by lukasheinrich** [comment]
maybe also related to #142 returning the axes as suggested there allows you to adjust the plot too. Even more advanced would be to follow the scikit-hep Plotting API style

https://scikit-learn.org/stable/developers/plotting.html

![image](https://user-images.githubusercontent.com/2318083/97803954-f812b580-1c4c-11eb-9fc8-5c0fa2fa9f99.png)


with 

```
class BrazilBandPlot(object):
    def __init__(self,ax,mutests,obs,exp,test_size):
        self.ax = ax
        self.obs = obs
        self.exp = exp
        self.mutests = mutests
        self.test_size = test_size

    def plot(self):
        self.ax.plot(self.mutests, self.obs, c='black')
        for idx, color in zip(range(5), 5 * ['black']):
            self.ax.plot(
                self.mutests, self.exp[idx], c=color, linestyle='dotted' if idx != 2 else 'dashed'
            )
        self.two_sigma_band =   self.ax.fill_between(self.mutests, self.exp[0], self.exp[-1], facecolor='yellow')
        self.ax.fill_between(self.mutests, self.exp[1], self.exp[-2], facecolor='green')
        
        self.ax.plot(self.mutests, [self.test_size] * len(self.mutests), c='red')
        self.ax.set_ylim(0, 1)

def plot_results(mutests, tests, test_size=0.05, ax = None):
    """Plot a series of hypothesis tests for various POI values."""
    ax = ax or plt.gca()
    cls_obs = np.array([test[0] for test in tests]).flatten()
    cls_exp = [np.array([test[1][i] for test in tests]).flatten() for i in range(5)]
    a = BrazilBandPlot(ax,mutests,cls_obs,cls_exp,test_size)
    a.plot()
    return a
```

**Comment by henryiii** [comment]
Not following too closely, but for matplotlib you should always return the Artists. You can get the parent Axes from the Artist, and you can get the Figure from an Axes. That's why all matplotlib plot functions return the created artists only.

**Comment by lukasheinrich** [comment]
in this case the entire plot is a collection of artists that are combined into a wrapper object `BrazilBandPlot` which holds artists `self. two_sigma_band`, `self.one_sigma_band`, etc. and the wrapper object is returned.

**Comment by alexander-held** [comment]
Closing this, as the original feature requested was implemented in #119. The callback idea can be tracked via #251, and will probably not make it into `v0.3` and then will be tracked via another issue that will be linked from there.

#264 will return figures from `visualize` function calls, allowing for further customization. Figures to me seem like the easier-to-work-with alternative to artists for this high level API. To my understanding, it would be a good idea to return artists from the lower-level figure creation implemented in `visualize.plot_result` and `visualize.plot_model`. These functions should be made more modular for that purpose.
